feat: Implement /eth/v1/node/peers/{peer_id}

NuttakitDW opened the pull request:
### What are you trying to achieve?

Resolve [Implement /eth/v1/node/peers/{peer_id} #222](https://github.com/ReamLabs/ream/issues/222)

### How was it implemented/fixed?

- **New route** – Added `GET /node/peers/{peer_id}` to align with the beacon-node API spec.  
- **Input validation** – `peer_id` is parsed with `libp2p::PeerId::from_str`; invalid IDs fail fast with `ApiError::BadRequest`.  
- **Response shape** – Returns JSON-wrapped `PeerData`  
  ```json
  {
    "peer_id": "<peer_id>",
    "enr": null,
    "last_seen_p2p_address": "<p2p_address>",
    "state": "connected",
    "direction": "outbound"
  }

- **Error handling** – Parsing and lookup errors map to `ApiError::{BadRequest, NotFound, InternalError}` and are logged via `tracing` for easier debugging.  
- **Stub data source** – Introduced `mock_fetch_peer` as a temporary stand-in; it supplies hard-coded peer info while we wire up the real `Network` integration.

#### Example request

```bash
curl -X GET -H "Accept: application/json" http://localhost:5052/eth/v1/node/peers/16Uiu2HAmUaM4YkWtwpu6P299nZuDwTCyUsrXeArCynSS1irSjXAM
```

<details> <summary>Sample response</summary>

```json
{
  "data": {
    "peer_id": "16Uiu2HAm1oEch6uXffoGZ32kPTiyjycfX9yDuBJSWtmagBSk9HTN",
    "enr": null,
    "last_seen_p2p_address": "/ip4/127.0.0.1/tcp/9000/p2p/16Uiu2HAm1oEch6uXffoGZ32kPTiyjycfX9yDuBJSWtmagBSk9HTN",
    "state": "connected",
    "direction": "outbound"
  }
}
```
</details> 

### To-Do

- [x] I have read [CONTRIBUTING.md](https://github.com/ReamLabs/ream/blob/master/CONTRIBUTING.md).
- [x] This PR uses [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Confirm that `PeerData` must be fetched directly from `p2p/src/network.rs::Network` (not from the database) for the `/node/peers/{peer_id}` endpoint.
- [x] Implement a read function (e.g., `fn cached_peer(&self, id: &PeerId) -> Option<CachedPeer>`) in `p2p/src/network.rs` to expose live peer data to the RPC layer.
- [x] Inject `Arc<Network>` into the RPC server (`start_server`) and register it with `App::app_data` so that handlers can call `network.cached_peer(...)`.

KolbyML commented:
I will only merge PR's that work, I won't merge PR's that return mock data

NuttakitDW commented:
@KolbyML Understood. This draft is only for design feedback. the mock data will be replaced. I’ll flip it to “Ready for review” once the real implementation is completed.

NuttakitDW commented:
Hi @KolbyML  this PR is ready for review
What’s in this PR

1. Network module

- Added peer-table with upsert + getter.
- Hooked all relevant SwarmEvents to keep the table in-sync.
- Added an integration test that walks a peer through connecting → disconnecting and asserts table state.

2. RPC
- Implemented GET /eth/v1/node/peers/{peer_id}.
- Handler pulls live info from the shared peer-table and returns spec-compliant JSON.

3. Wiring
- Exposed peer_table() on Network.
- Passed the table into start_server (app_data) so the RPC layer can access it.

KolbyML commented on bin/ream/src/main.rs:
```diff
@@ -101,22 +101,26 @@ async fn main() {
             info!("Database Initialization completed");
 
-            let http_future = start_server(server_config, ream_db);
+            let mut network = match Network::init(async_executor, &network_config).await {
+                Ok(net) => net,
+                Err(e) => {
+                    error!("Failed to initialise network: {e}");
+                    return;
+                }
+            };
```
```suggestion
            let mut network = Network::init(async_executor, &network_config).await.expect("Failed to initialize P2P Network");
```

KolbyML commented on crates/networking/p2p/src/constants.rs:
```diff
@@ -4,3 +4,10 @@ use alloy_primitives::{aliases::B32, fixed_bytes};
 pub const MAX_PAYLOAD_SIZE: u64 = 10485760;
 pub const MESSAGE_DOMAIN_VALID_SNAPPY: B32 = fixed_bytes!("0x01000000");
 pub const MESSAGE_DOMAIN_INVALID_SNAPPY: B32 = fixed_bytes!("0x00000000");
+
+pub const STATE_CONNECTED: &str = "connected";
+pub const STATE_CONNECTING: &str = "connecting";
+pub const STATE_DISCONNECTED: &str = "disconnected";
+pub const STATE_DISCONNECTING: &str = "disconnecting";
+pub const DIR_INBOUND: &str = "inbound";
+pub const DIR_OUTBOUND: &str = "outbound";
```
These constants should be in the p2p crate, because they are only used in that crate

also please avoid shorthand in `DIR_INBOUND` for reference

NuttakitDW replied:
Copy that. I’ve removed these constants since I’ll be using enums for the state and direction fields in the CachedPeer struct.

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -28,20 +29,32 @@ use libp2p::{
+pub struct CachedPeer {
+    pub peer_id: PeerId,
+    pub last_seen_p2p_address: Option<Multiaddr>,
+    pub state: &'static str, // "connected" | "connecting" | "disconnected" | "disconnecting"
+    pub direction: &'static str, // "inbound" | "outbound" | ""
```
Can we use enums here instead of static str's

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -406,9 +517,102 @@ mod tests {
 
+    #[test]
+    fn peer_id_from_enr_matches_libp2p() {
+        let lp_kp = Keypair::generate_secp256k1();
```
can you avoid shorthands in the tests you added as well, I won't point out every instance, but generally shorthands aren't readable for new contributors so we want to avoid them, to make everything as readable as possible

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -406,9 +517,102 @@ mod tests {
+    fn insert_then_read_returns_snapshot() {
+        init_network_spec();
+
+        let rt = tokio::runtime::Runtime::new().unwrap();
```
```suggestion
        let tokio_runtime = Runtime::new().unwrap();
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -406,9 +517,102 @@ mod tests {
+                .unwrap()
+        });
+
+        let peer_id = libp2p::PeerId::random();
```
```suggestion
        let peer_id = PeerId::random();
```

everywhere else as well

KolbyML commented on crates/rpc/Cargo.toml:
```diff
@@ -17,6 +17,7 @@ ream-light-client.workspace = true
 ream-network-spec.workspace = true
 ream-node.workspace = true
 ream-storage.workspace = true
+ream-p2p.workspace = true
```
put this in the list alphabetically

KolbyML commented on crates/rpc/Cargo.toml:
```diff
@@ -31,3 +32,4 @@ ssz_types.workspace = true
 thiserror.workspace = true
 tracing.workspace = true
 tree_hash.workspace = true
+libp2p.workspace = true
```
put this in the list alphabetically

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -366,9 +468,18 @@ mod tests {
         gossipsub::{configurations::GossipsubConfig, topics::GossipTopicKind},
     };
 
+    static INIT_NET_SPEC: Once = Once::new();
```
```suggestion
    static HAS_NETWORK_SPEC_BEEN_INITIALIZED: Once = Once::new();
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -366,9 +468,18 @@ mod tests {
 
+    static INIT_NET_SPEC: Once = Once::new();
+
+    fn init_network_spec() {
```
```suggestion
    fn initialize_network_spec() {
```

KolbyML reviewed (commented):
I left an early review, I will give another review after as this PR is a bit to go through. Anyways don't be scared to ping me again, sorry for the delay in reviewing this :saluting_face: looks pretty good

NuttakitDW commented:
Thanks @KolbyML !! Good to hear that 😆

KolbyML commented:
hi @NuttakitDW to rebase you should do

```
git fetch upstream
git rebase upstream/master
```

NuttakitDW commented:
ok copy that @KolbyML

NuttakitDW commented:
Hi @KolbyML 
I’ve addressed all your feedback and updated the test names and variables to avoid shorthands for better readability. Let me know if there’s anything else you'd like adjusted. Thanks again for the thorough review!

KolbyML commented on crates/rpc/src/handlers/peers.rs:
```diff
@@ -0,0 +1,55 @@
+
+use crate::types::{errors::ApiError, response::DataResponse};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PeerData {
+    peer_id: String,
+    enr: Option<String>,
+    last_seen_p2p_address: String,
+    state: String,
+    direction: String,
+}
+
+impl From<&CachedPeer> for PeerData {
+    fn from(p: &CachedPeer) -> Self {
+        Self {
+            peer_id: p.peer_id.to_string(),
+            enr: p.enr.as_ref().map(|e| e.to_base64()),
+            last_seen_p2p_address: p
+                .last_seen_p2p_address
+                .as_ref()
+                .map(|m| m.to_string())
+                .unwrap_or_default(),
+            state: p.state.as_str().to_string(),
+            direction: p.direction.as_str().to_string(),
+        }
+    }
+}
```
```suggestion
```

We should just use CachedPeer directly

KolbyML commented on crates/rpc/src/handlers/peers.rs:
```diff
@@ -0,0 +1,55 @@
+/// GET /eth/v1/node/peers/{peer_id}
+#[get("/node/peers/{peer_id}")]
+pub async fn get_peer(
+    table: Data<PeerTable>,
```
```suggestion
    peer_table: Data<PeerTable>,
```

KolbyML commented on crates/rpc/src/handlers/peers.rs:
```diff
@@ -0,0 +1,55 @@
+    let peer_id = PeerId::from_str(&id_raw)
+        .map_err(|_| ApiError::BadRequest(format!("Invalid peer ID: {id_raw}")))?;
+
+    let snap = table
```
I don't know what snap means

KolbyML commented on crates/rpc/src/lib.rs:
```diff
@@ -11,20 +12,27 @@ pub mod routes;
             .wrap(middleware::Logger::default())
             .app_data(stop_handle)
             .app_data(Data::new(db.clone()))
+            .app_data(peer_table.clone())
```
```suggestion
            .app_data(Data::new(peer_table.clone()))
```

KolbyML commented on crates/rpc/src/lib.rs:
```diff
@@ -11,20 +12,27 @@ pub mod routes;
 
     let server = HttpServer::new(move || {
         let stop_handle = stop_handle.clone();
+        let peer_table = peer_table_data.clone();
```
```suggestion
```

KolbyML commented on crates/rpc/src/lib.rs:
```diff
@@ -11,20 +12,27 @@ pub mod routes;
     );
     // create the stop handle container
     let stop_handle = Data::new(StopHandle::default());
+    let peer_table_data = Data::new(peer_table);
```
```suggestion
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -239,6 +339,50 @@ impl Network {
+            SwarmEvent::ConnectionEstablished {
+                peer_id, endpoint, ..
+            } => {
+                let (direction, addr) = match &endpoint {
+                    ConnectedPoint::Dialer { address, .. } => {
+                        (Direction::Outbound, Some(address.clone()))
+                    }
+                    ConnectedPoint::Listener { send_back_addr, .. } => {
+                        (Direction::Inbound, Some(send_back_addr.clone()))
+                    }
+                };
+
+                self.upsert_peer(peer_id, addr, ConnectionState::Connected, direction, None);
```
```suggestion
                let (direction, address) = match &endpoint {
                    ConnectedPoint::Dialer { address, .. } => {
                        (Direction::Outbound, Some(address.clone()))
                    }
                    ConnectedPoint::Listener { send_back_addr, .. } => {
                        (Direction::Inbound, Some(send_back_addr.clone()))
                    }
                };

                self.upsert_peer(peer_id, address, ConnectionState::Connected, direction, None);
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -219,6 +269,56 @@ impl Network {
+        }
+    }
+
+    fn upsert_peer(
+        &mut self,
+        peer_id: PeerId,
+        addr: Option<Multiaddr>,
+        state: ConnectionState,
+        direction: Direction,
+        enr: Option<Enr>,
+    ) {
+        let mut table = self.peer_table.write();
+        table
+            .entry(peer_id)
+            .and_modify(|row| {
+                if addr.is_some() {
+                    row.last_seen_p2p_address = addr.clone();
+                }
+                row.state = state;
+                row.direction = direction;
+                if enr.is_some() {
+                    row.enr = enr.clone();
+                }
+            })
+            .or_insert(CachedPeer {
+                peer_id,
+                last_seen_p2p_address: addr,
+                state,
+                direction,
+                enr,
+            });
+    }
```
```suggestion
    fn upsert_peer(
        &mut self,
        peer_id: PeerId,
        address: Option<Multiaddr>,
        state: ConnectionState,
        direction: Direction,
        enr: Option<Enr>,
    ) {
        self.peer_table.write()
            .entry(peer_id)
            .and_modify(|row| {
                if addr.is_some() {
                    row.last_seen_p2p_address = address.clone();
                }
                row.state = state;
                row.direction = direction;
                if enr.is_some() {
                    row.enr = enr.clone();
                }
            })
            .or_insert(CachedPeer {
                peer_id,
                last_seen_p2p_address: address,
                state,
                direction,
                enr,
            });
    }
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -28,9 +29,9 @@ use libp2p::{
     yamux,
 };
-use libp2p_identity::{Keypair, PublicKey, secp256k1};
+use libp2p_identity::{Keypair, PublicKey, secp256k1, secp256k1::PublicKey as LibSecpPk};
```
```suggestion
use libp2p_identity::{Keypair, PublicKey, secp256k1, secp256k1::PublicKey as Secp256k1PublicKey};
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -219,6 +269,56 @@ impl Network {
+        self.peer_table.read().get(id).cloned()
+    }
+
+    fn peer_id_from_enr(enr: &Enr) -> Option<PeerId> {
+        match enr.public_key() {
+            CombinedPublicKey::Secp256k1(pk) => {
+                let encoded = pk.to_encoded_point(true);
+                let lib_secp = LibSecpPk::try_from_bytes(encoded.as_bytes()).ok()?;
+                let lib_pk: PublicKey = lib_secp.into();
+                Some(PeerId::from_public_key(&lib_pk))
+            }
+            _ => None,
+        }
+    }
```
```suggestion
    fn peer_id_from_enr(enr: &Enr) -> anyhow::Result<Option<PeerId>> {
       Ok(match enr.public_key() {
            CombinedPublicKey::Secp256k1(public_key) => {
                let encoded_public_key = public_key.to_encoded_point(true);
                let public_key = Secp256k1PublicKey::try_from_bytes(encoded_public_key.as_bytes())?.into();
                 Some(PeerId::from_public_key(&public_key))
            }
            _ => None,
        })
    }
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -480,4 +750,84 @@ mod tests {
+            .expect("peer-table not updated in time");
+        });
+
+        let row1 = network1
+            .cached_peer(&peer_id_network2)
+            .expect("network1 row exists");
+        let row2 = network2
```
row what? peer_from_network_1? What is a row, the variable name should be more informative

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -480,4 +750,84 @@ mod tests {
+
+        tokio_runtime.block_on(async {
+            let network1_poll_task = async {
+                while let Some(ev) = network1.swarm.next().await {
+                    network1.parse_swarm_event(ev);
```
what is an ev?

NuttakitDW replied:
ev is event. I will change it asap.

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -411,10 +577,114 @@ mod tests {
+                .unwrap()
+        });
+
+        let random_id = PeerId::random();
```
```suggestion
        let peer_id = PeerId::random();
```

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -411,10 +577,114 @@ mod tests {
+        });
+
+        let peer_id = PeerId::random();
+        let addr: libp2p::Multiaddr = "/ip4/1.2.3.4/tcp/9000".parse().unwrap();
```
```suggestion
        let addr: Multiaddr = "/ip4/1.2.3.4/tcp/9000".parse().unwrap();
```

KolbyML reviewed (commented):
Here is some feedback

NuttakitDW commented:
thanks @KolbyML I will look into it.

NuttakitDW commented:
oh sorry @KolbyML my bad habit. 
I just like to commit every time I make a small change base on my todo list.
I will fix that from now on if you want.

NuttakitDW commented:
done! @KolbyML

NuttakitDW commented:
I forgot to run lint 🥶
Now it should be ok.

KolbyML reviewed (commented):
I forgot to hit submit on this

NuttakitDW commented on crates/networking/p2p/Cargo.toml:
```diff
@@ -18,6 +18,10 @@ enr.workspace = true
+k256 = { version = "0.13", default-features = false, features = [
+    "ecdsa",
+    "arithmetic",
+] }
```
ship it @KolbyML

KolbyML commented on crates/networking/p2p/src/network.rs:
```diff
@@ -642,4 +913,84 @@ mod tests {
+            ))
+            .unwrap();
+
+        let address: Multiaddr = "/ip4/127.0.0.1/tcp/9300".parse().unwrap();
+        network2.swarm.dial(address).unwrap();
+
+        let peer_id_network1 = network1.peer_id();
```
```suggestion
        let address: Multiaddr = "/ip4/127.0.0.1/tcp/9300".parse().unwrap();
        network_2.swarm.dial(address).unwrap();

        let peer_id_network_1 = network_1.peer_id();
```

we sure we have underscores between these to be consistent, check your PR for other cases this happens

KolbyML reviewed (approved):
:shipit: looks great thank you for your contribution :rocket:, after these concerns are resolved the PR should be good to merge

KolbyML commented on Cargo.toml:
```diff
@@ -60,7 +69,24 @@ hashbrown = "0.15.3"
 jsonwebtoken = "9.3.1"
 kzg = { git = "https://github.com/grandinetech/rust-kzg" }
-libp2p = { version = "0.55", default-features = false, features = ["identify", "yamux", "noise", "dns", "tcp", "tokio", "plaintext", "secp256k1", "macros", "ecdsa", "metrics", "quic", "upnp", "gossipsub", "ping"] }
+libp2p = { version = "0.55", default-features = false, features = [
+    "identify",
+    "yamux",
+    "noise",
+    "dns",
+    "serde",
+    "tcp",
+    "tokio",
+    "plaintext",
+    "secp256k1",
+    "macros",
+    "ecdsa",
+    "metrics",
+    "quic",
+    "upnp",
+    "gossipsub",
+    "ping",
+] }
```
> okie do you want me to check every dpecdencies in carog.toml to make it to inlined? eg.
> 
> ```
>  tokio = { version = "1.42", features = [
>     "rt",
>     "rt-multi-thread",
>     "sync",
>     "signal",
>     "time",
>     "macros",
> ] }
> ```

this should be 1 line like it used to be

KolbyML commented:
@NuttakitDW you gotta fix ci, make sure you have the latest version of cargo-sort installed, it might be out of date

NuttakitDW commented:
My CI is fixed!
I'm ready to go 🚀
